src/python/pants/backend/project_info/rules/dependencies.py

```diff
@@ -52,22 +53,38 @@ class Dependencies(Goal):
     console: Console, addresses: Addresses, options: DependenciesOptions,
 ) -> Dependencies:
     address_strings: Set[str] = set()
+    third_party_requirements: Set[str] = set()
+
     if options.values.transitive:
         transitive_targets = await Get[TransitiveHydratedTargets](Addresses, addresses)
-        transitive_dependencies = transitive_targets.closure - set(transitive_targets.roots)
-        address_strings.update(
-            hydrated_target.address.spec for hydrated_target in transitive_dependencies
+        hydrated_targets = HydratedTargets(
+            transitive_targets.closure - set(transitive_targets.roots)
         )
     else:
         hydrated_targets = await Get[HydratedTargets](Addresses, addresses)
-        address_strings.update(
-            dep.spec for hydrated_target in hydrated_targets for dep in hydrated_target.dependencies
-        )
-
-    with options.line_oriented(console) as print_stdout:
-        for address in sorted(address_strings):
-            print_stdout(address)
 
+    for target in hydrated_targets:
+        address_string = target.address.spec
+        if address_string.startswith("3rdparty"):
+            if isinstance(target.adaptor, PythonRequirementLibraryAdaptor):
+                third_party_requirements.update(
+                    str(requirement.requirement) for requirement in target.adaptor.requirements
+                )
+            # TODO: Support jvm third party deps when there is some sort of JarLibraryAdaptor.
+        else:
+            address_strings.update(
+                hydrated_target.address.spec for hydrated_target in hydrated_targets
+            )
+
+    if options.values.type in [DependencyType.SOURCE, DependencyType.SOURCE_AND_THIRD_PARTY]:
+        with options.line_oriented(console) as print_stdout:
```

Comment:
This gets repeated again. Intead, you can put the `if options.values.type in [DependencyType.SOURCE, DependencyType.SOURCE_AND_THIRD_PARTY]` code and `if options.values.type in [DependencyType.THIRD_PARTY, DependencyType.SOURCE_AND_THIRD_PARTY]` within the same `with` statement.

--

Also, if you take the above recommendation to only add to `address_strings` and `third_party_requirements` when relevant, then you can remove the if statements here and always unconditionally print the addresses and the requirements. Those will be empty if `--type` says to not output them.

What should the ordering be for `--type=source-and-3rdparty`? I think V1 is unordered? Maybe merge both and have it be sorted.

Reply:
> What should the ordering be for --type=source-and-3rdparty? I think V1 is unordered? Maybe merge both and have it be sorted.

For me I prefer to split these so you get output that looks like this:
```
[...]
src/python/pants/util:strutil
src/python/pants/util:tarutil
src/python/pants/util:xml_parser
src/python/pants:entry_point
src/python/pants:version
src/python/pants:version-resource
src/rust/engine/process_execution/bazel_protos:bazel_protos
src/rust/engine:engine
Markdown==2.1.1
PyYAML==5.1.2
Pygments==2.3.1
ansicolors==1.1.8
beautifulsoup4<4.7,>=4.6.0
cffi==1.13.2
contextlib2==0.5.5
[...]
```

Rather than mixing all those 3rd party strings into the source target addresses, eg, the output of V1:

```
[...]
contrib/python/src/python/pants/contrib/python/checks/tasks:resources
contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle:checkstyle
contrib/python/src/python/pants/contrib/python/checks/checker:checker
contrib/python/src/python/pants/contrib/python/checks/checker/3rdparty:pycodestyle
pycodestyle==2.4.0
contrib/python/src/python/pants/contrib/python/checks/checker/3rdparty:requirements.txt
contrib/python/src/python/pants/contrib/python/checks/checker/3rdparty:pyflakes
pyflakes==2.1.1
contrib/python/src/python/pants/contrib/python/checks/checker/3rdparty:six
six<2,>=1.9.0
contrib/scalajs/src/python/pants/contrib/scalajs:plugin
contrib/scalajs/src/python/pants/contrib/scalajs/subsystems:subsystems
contrib/scalajs/src/python/pants/contrib/scalajs/targets:targets
[...]
```
Sorting those strings helps some, but you'll still get some requirements strings, then a bunch of targets under `contrib`, then some more requirement strings, then a bunch of targets under `src` and so on.

Reply:
Oh, the first is indeed quite nice! Can you please sort within both those collections, though?

Reply:
They are sorted within both collections, But it's python lexical sort, so PyYAML comes before ansicolors (capitalization sucks)